### PR TITLE
Add Documentation

### PR DESCRIPTION
**Pull request by adrianchiris**
- Update README
- Add CONTRIBUTING doc

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,19 +1,158 @@
...
+
+1. User (cluster administrator) defines a set of named IP Pools to be used for IP allocation
+of container interfaces via Kubernetes ConfigMap (more information in [Configuration](#configuration) section)
+
+_Example_:
+
+```json
+{
+    "pools":  {
+        "my-pool": {"subnet": "192.188.0.0/16", "perNodeBlockSize": 24, "gateway": “192.168.0.1”},
```
Comment (ykulazhenkov) [inline README.md:59]: Invalid quotes the for gateway option.

File: README.md [hunk 2]
```diff
@@ -1,19 +1,161 @@
...
+[![CodeQL](https://github.com/Mellanox/nvidia-k8s-ipam/actions/workflows/codeql.yml/badge.svg)](https://github.com/Mellanox/nvidia-k8s-ipam/actions/workflows/codeql.yml)
+[![Image push](https://github.com/Mellanox/nvidia-k8s-ipam/actions/workflows/image-push-main.yml/badge.svg?event=push)](https://github.com/Mellanox/nvidia-k8s-ipam/actions/workflows/image-push-main.yml)
 
-### Deploy IPAM
+An IP Address Management (IPAM) CNI plugin designed to operate in a Kubernetes environment.
+This Plugins allows to assign IP addresses dynamically across the cluster while keeping speed
+and performance in mind.
 
-_NOTE: This command will deploy latest dev build_
+IP subnets are defined by the user as named _IP Pools_, then for each IP Pool a unique _IP Block_ is assinged
```
Comment (moshe010) [inline README.md:14]: s/assinged/assigned

File: README.md [hunk 3]
```diff
@@ -1,19 +1,161 @@
...
+A node agent that performs initial setup and installation of nv-ipam CNI plugin.
+
+### nv-ipam
+
+An IPAM CNI plugin that allocates IPs for a given interface out of the defined IP Pool as provided
+via CNI configuration.
+
+IPs are allocated out of the provided IP Block assigned by ipam-controller for the node.
+To determine the cluster unique IP Block for the defined IP Pool, nv-ipam CNI queries K8s API
+for the Node object and extracts IP Block infromation from node annotation.
```
Comment (moshe010) [inline README.md:47]: s/infromation/information

File: README.md [hunk 4]
```diff
@@ -23,20 +165,101 @@ data:
...
+  * `subnet`: IP Subnet of the pool
+  * `gateway`: Gateway IP of the subnet
+  * `perNodeBlockSize`: the number of IPs of IP Blocks allocated to Nodes.
+* `nodeSelector`: a map<string, string> of node selector labels, only nodes that match the provided labels will get assigned IP Blocks for the defined pools
+
+> __Notes:__
+>
+> * pool name is composed of alphanumeric letters separated by dots(`.`) undersocres(`_`) or hyphens(`-`)
+> * `perNodeBlockSize` minimum size is 2
+> * `subnet` must be large enough to accomodate at least one `perNodeBlockSize` block of IPs
```
Comment (moshe010) [inline README.md:184]: s/accomodate/accommodate

File: README.md [hunk 5]
```diff
@@ -23,20 +165,101 @@ data:
   config: |
     {
       "pools": {
-        "pool1": {"subnet": "192.168.0.0/16", "perNodeBlockSize": 100 , "gateway": "192.168.0.1"},
-        "pool2": {"subnet": "172.16.0.0/16", "perNodeBlockSize": 50 , "gateway": "172.16.0.1"}
+        "my-pool": {"subnet": "192.168.0.0/16", "perNodeBlockSize": 100 , "gateway": "192.168.0.1"},
```
Comment (moshe010) [inline README.md:168]: remove last , so it will be valid json

File: README.md (hunk 1 above)
Comment (moshe010) [reply README.md:62]: Also remove last , so it will be valid json

File: README.md [hunk 6]
```diff
@@ -1,19 +1,161 @@
...
+                The name of the ConfigMap which holds controller configuration (default "nvidia-k8s-ipam-config")
+      --config-namespace string                                                                                                                                                                                        
+                The name of the namespace where ConfigMap with controller configuration exist (default "kube-system")
+      --health-probe-bind-address string                                                                                                                                                                               
+                The address the probe endpoint binds to. (default ":8081")
+      --kubeconfig string                                                                                                                                                                                              
+                Paths to a kubeconfig. Only required if out-of-cluster.
+      --leader-elect
```
Comment (moshe010) [inline README.md:148]: I thought we don't support it. Maybe we should document only the things that will work

File: README.md [hunk 7]
```diff
@@ -23,20 +165,101 @@ data:
...
+      --cni-conf-dir string                   CNI config directory (default "/host/etc/cni/net.d")
+  -h, --help                                  show help message and quit
+      --host-local-bin-file string            host-local binary file path (default "/host-local")
+      --nv-ipam-bin-file string               nv-ipam binary file path (default "/nv-ipam")
+      --nv-ipam-cni-data-dir string           nv-ipam CNI data directory (default "/host/var/lib/cni/nv-ipam")
+      --nv-ipam-cni-data-dir-host string      nv-ipam CNI data directory on host (default "/var/lib/cni/nv-ipam")
+      --nv-ipam-kubeconfig-file-host string   kubeconfig for nv-ipam (default "/etc/cni/net.d/nv-ipam.d/nv-ipam.kubeconfig")
+      --nv-ipam-log-file string               nv-ipam log file (default "/var/log/nv-ipam-cni.log")
+      --nv-ipam-log-level string              nv-ipam log level (default "info")
+      --skip-host-local-binary-copy           skip host-loca binary file copy
```
Comment (moshe010) [inline README.md:201]: missing 
--skip-nv-ipam-binary-copy              skip nv-ipam binary file copy

File: README.md [hunk 8]
```diff
@@ -23,20 +165,101 @@ data:
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: nvidia-k8s-ipam-config
+  namespace: kube-system
+data:
+  config: |
+    {
+      "pools": {
+        "my-pool": {"subnet": "192.168.0.0/16", "perNodeBlockSize": 100 , "gateway": "192.168.0.1"},
```
Comment (moshe010) [inline README.md:251]: remove last , so it will be valid json

**Review by moshe010 — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (ykulazhenkov) [reply README.md:148]: In the code it is supported. 
I added [1] extra config option to configure namespace for leader election process. Also enabled leader election by default to protect from the situation when a user decided to manually scale the deployment. 

[1] https://github.com/Mellanox/nvidia-k8s-ipam/pull/19

File: README.md (hunk 1 above)
Comment (adrianchiris) [reply README.md:62]: ack, this was copied from our doc, i have fixed there as well.

File: README.md [hunk 9]
```diff
@@ -51,5 +274,53 @@ cat > /etc/cni/net.d/10-mynet.conf <<EOF
...
+
+> _Note:_ It is recommended to increase log level to `"debug"` when debugging CNI issues.
+> This can be done by either editing CNI configuration (e.g editing NetworkAttachmentDefinition)
+> or locally on a node by editing nv-ipam config file (default: `/etc/init.d/nv-ipam.d/nv-ipam.conf`
+
+## Limitations
+
+* Deleting an IP Pool from config map while there are pods scheduled on nodes with IPs from deleted Pool, deleting these pods will fail (CNI CMD DEL fails)
+* Before removing a node from cluster, drain all workloads to ensure proper cleanup of IPs on node.
+* IP Block allocated to a node with Gateway IP in its range will have one less IP than what defined in perNodeblockSize, deployers should take this into account.
```
Comment (moshe010) [inline README.md:319]: s/perNodeblockSize/perNodeBlockSize

File: README.md (hunk 6 above)
Comment (adrianchiris) [reply README.md:148]: These are the supported flags by the binary, should i remove or keep ?

either way #19 would need to update this file as well once we merge it.

File: README.md (hunk 8 above)
Comment (adrianchiris) [reply README.md:251]: done

**Comment by adrianchiris** [comment]
@ykulazhenkov @moshe010 PTAL

lets get this one in

**Review by moshe010 — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -51,5 +273,54 @@ cat > /etc/cni/net.d/10-mynet.conf <<EOF
...
+* Deleting an IP Pool from config map while there are pods scheduled on nodes with IPs from deleted Pool, deleting these pods will fail (CNI CMD DEL fails)
+* Before removing a node from cluster, drain all workloads to ensure proper cleanup of IPs on node.
+* IP Block allocated to a node with Gateway IP in its range will have one less IP than what defined in perNodeBlockSize, deployers should take this into account.
+* IPv6 not supported
+* Allocating Multiple IPs per interface is not supported
+* Defining multiple IP Pools while supported, was not thoroughly testing
+
+## Contributing
+
+We welcome your feedback and contributions to this project. Please see the [CONTRIBUTING.md](https://github.com/Mellanox/nvidia-k8s-ipam/blob/main/CONTRIBUTING.md)
```
Comment (ykulazhenkov) [inline README.md:325]: Do we want to use relative link here? `[CONTRIBUTING.md](CONTRIBUTING.md)`

**Review by ykulazhenkov — APPROVED** [review]

**Pull request merged**